Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/PreferReadOnlyParametersAnalyzer.cs

```diff
@@ -0,0 +1,108 @@
+
+namespace Philips.CodeAnalysis.MaintainabilityAnalyzers.Maintainability
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class PreferReadOnlyParametersAnalyzer : SingleDiagnosticAnalyzer<MethodDeclarationSyntax, PreferReadOnlyParametersNodeAction>
+	{
+		private const string Title = @"Prefer readonly parameters";
+		public const string MessageFormat = @"Make parameter '{0}' a readonly type";
+		private const string Description = @"Method parameters that are not written to, can be made readonly";
+		public PreferReadOnlyParametersAnalyzer()
+			: base(DiagnosticId.PreferReadOnlyParameters, Title, MessageFormat, Description, Categories.Maintainability)
+		{ }
+	}
+
+	public class PreferReadOnlyParametersNodeAction : SyntaxNodeAction<MethodDeclarationSyntax>
+	{
+		private static readonly List<string> ReadWriteCollections = new() { "System.Array", "System.Collections.Generic.List", "System.Span" };
+		private readonly NamespaceIgnoringComparer _comparer = new();
+
+		public override void Analyze()
+		{
+			var parameters = Node.ParameterList.Parameters;
+			if (!parameters.Any())
+			{
+				return;
+			}
+
+			var aliases = Helper.GetUsingAliases(Node);
+			var collections = parameters.Where(p => IsReadWriteCollection(p.Type, aliases));
+			if (!collections.Any())
+			{
+				return;
+			}
+			var accesses = Node.DescendantNodes().OfType<MemberAccessExpressionSyntax>().ToList();
+			var setters = Node.DescendantNodes()
+				.OfType<AssignmentExpressionSyntax>()
+				.Select(ass => ass.Left as ElementAccessExpressionSyntax).ToList();
+			var invocations = Node.DescendantNodes().OfType<InvocationExpressionSyntax>().ToList();
+			foreach (ParameterSyntax collectionParameter in collections)
+			{
+				var parameterName = collectionParameter.Identifier.Text;
+				// Calling member that will change the type...
+				// Or using the setter of the indexer (example: list[8] = 42) ...
+				// Or invocations that require the type
+				if (
+					!accesses.Any(acc => IsCallingParameter(acc.Expression, parameterName) && IsModifyingMember(acc.Name)) &&
+					!setters.Any(element => element != null && IsCallingParameter(element.Expression, parameterName)) &&
+					!invocations.Any(voc => RequiresReadWrite(voc, parameterName)))
+				{
+					var location = collectionParameter.Type.GetLocation();
+					ReportDiagnostic(location, collectionParameter.Identifier.Text);
+				}
+			}
+		}
+
+		private bool IsReadWriteCollection(TypeSyntax type, IReadOnlyDictionary<string, string> aliases)
+		{
+			if (type == null)
+			{
+				return false;
+			}
+			var fullName = type.GetFullName(aliases);
+			return ReadWriteCollections.Any(col => _comparer.Compare(fullName, col) == 0);
+		}
+
+		private bool IsCallingParameter(ExpressionSyntax expression, string name)
+		{
+			return expression is IdentifierNameSyntax identifierName && _comparer.Compare(identifierName.Identifier.Text, name) == 0;
+		}
+
+		private static bool IsModifyingMember(SimpleNameSyntax name)
```

Comment:
It feels preferable to attack this in the opposite direction. E.g., if it only uses [], then IEnumerable, if it only uses [] and Count, then IReadonly, if it only uses Count, then int.

Reply:
Would be great for the CodeFixer to be able to pick the correct type / interface

Reply:
Created an issue for this "fine tuning" of the proposed type.